File: working_on/17677.py
# 뉴스 클러스터링
# 50분 소요! 1트에 성공 :)

import string


def cleansing(str):
    # 1. 두 글자씩 끊어서 다중집합으로 정제
    # 2. 영문자만 유효
    # 3. 대소문자 구분 X

    # 모두 소문자로 바꿔준다
    s = list(str.lower())
    c = []

    alphabet = list(string.ascii_lowercase)
    for i in range(len(s)):
        # 두 글자씩 자르기
        couple = s[i : i + 2]
        if len(couple) == 2:
            if couple[0] in alphabet and couple[1] in alphabet:
                c.append("".join(couple))

    return c


def solution(str1, str2):
    a = cleansing(str1)
    b = cleansing(str2)
    # print(a, b)

    # 둘 다 공집합인 경우
    if len(a) == 0 and len(b) == 0:
        return int(1 * 65536)

    # A의 길이가 더 크다고 가정
    # if len(a) <= len(b):
    #     a, b = b, a

    inter = []

    # 교집합의 길이 구하기
    for item in set(a):
        if item in b:
            num_item = min(a.count(item), b.count(item))
            for i in range(num_item):
                inter.append(item)

    union = []

    # 합집합의 길이 구하기
    for item in set(a + b):
        num_item = max(a.count(item), b.count(item))
        print(num_item)
        for i in range(num_item):
            union.append(item)

    # J(A,B) = A,B 교집합 / A,B 합집합
    return int((len(inter) / len(union)) * 65536)


str1, str2 = "FRANCE", "french"
# str1, str2 = "handshake", "shake hands"
# str1, str2 = "aa1+aa2", "AAAA12"
# print(cleansing(str1))
# print(cleansing(str2))
print(solution(str1, str2))

File: programmers/lvl_2/17677.js
// 뉴스 클러스터링

function solution(str1, str2) {
  const a = cleansing(str1);
  const b = cleansing(str2);

  console.log(a, b);

  if (a.length == 0 && b.length == 0) {
    return Math.floor(1 * 65536);
  }

  let inter = [];

  let a_set = [...new Set(a)];

  console.log("흠", a_set);

  for (item of a_set) {
    if (b.includes(item)) {
      n = Math.min(
        a.filter((x) => x == item).length,
        b.filter((x) => x == item).length
      );
      for (i = 0; i < n; i++) {
        inter.push(item);
      }
    }
  }

  console.log("교집합: ", inter);

  let union = [];

  console.log(a.concat(b));

  let u_set = [...new Set(a.concat(b))];
  console.log("시바", u_set);

  for (item of u_set) {
    n = Math.max(
      a.filter((x) => x == item).length,
      b.filter((x) => x == item).length
    );
    for (i = 0; i < n; i++) {
      union.push(item);
    }
  }

  return Math.floor((inter.length / union.length) * 65536);
}

function cleansing(str) {
  const s = Array.from(str.toUpperCase());
  const alpha = Array.from(Array(26)).map((e, i) => i + 65);
  const alphabet = alpha.map((x) => String.fromCharCode(x));
  //   console.log(alphabet);
  let c = [];
  for (i = 0; i < s.length; i++) {
    let couple = s.slice(i, i + 2);
    // console.log(couple);
    if (couple.length == 2) {
      if (alphabet.includes(couple[0]) && alphabet.includes(couple[1])) {
        c.push(couple[0] + couple[1]);
      }
    }
  }
  return c;
}

str1 = "FRANCE";
str2 = "french";
// console.log(cleansing(str2));
console.log(solution(str1, str2));
